src/soroban_rpc.ts

```diff
@@ -162,16 +162,77 @@ export namespace SorobanRpc {
     retval: xdr.ScVal;
   }
 
-  export interface SimulateTransactionResponse {
+  /**
+   * Simplifies {@link RawSimulateTransactionResponse} into separate interfaces
+   * based on status:
+   *   - on success, this includes all fields, though `result` is only present
+   *     if an invocation was simulated (since otherwise there's nothing to
+   *     "resultify")
+   *   - if there was an expiration error, this includes error and restoration
+   *     fields
+   *   - for all other errors, this only includes error fields
+   *
+   * @see https://soroban.stellar.org/api/methods/simulateTransaction#returns
+   */
+  export type SimulateTransactionResponse =
+    | SimulateTransactionSuccessResponse
+    | SimulateTransactionErrorResponse;
+
+  interface BaseSimulateTransactionResponse {
+    /** always present: the JSON-RPC request ID */
     id: string;
-    error?: string;
-    transactionData: SorobanDataBuilder;
+
+    /** always present: the LCL known to the server when responding */
+    latestLedger: number;
+
+    /**
+     * The field is always present, but may be empty in cases where:
+     *   - you didn't simulate an invocation,
+     *   - there were no events, or
+     *   - you are communicating with an RPC server w/o diagnostic events on
+     */
     events: xdr.DiagnosticEvent[];
+  }
+
+  /** Includes simplified fields only present on success. */
+  export interface SimulateTransactionSuccessResponse
+    extends BaseSimulateTransactionResponse {
+    transactionData: SorobanDataBuilder;
     minResourceFee: string;
-    // only present if error isn't
-    result?: SimulateHostFunctionResult;
-    latestLedger: number;
     cost: Cost;
+
+    /** present only for invocation simulation */
+    result?: SimulateHostFunctionResult;
+
+    /**
+     * Indicates that a restoration is necessary prior to submission.
+     *
+     * In other words, seeing a restoration preamble means that your invocation
+     * was executed AS IF the required ledger entries were present, and this
+     * field includes information about what you need to restore for the
+     * simulation to succeed.
+     */
+    restore?: {
```

Comment:
Why rename this from `restorePreamble`? The other name seems fine and less confusing when comparing with the docs.